content/en/blog/_posts/qos-class-resources.md

```diff
@@ -0,0 +1,305 @@
+functionality gradually, step-by-step. The goal is to make discussions easier
+and also get feedback and learn on the way.
+
+#### Phase 1
+
+The goal of the first phase is to enable a bare minimum environment for users
+to use and experiment with QoS-class resources in Kubernetes:
+
+- extend the CRI protocol to allow QoS-class resource assignment and updates
+- implement pod annotations as an initial user interface in Kubernetes
+
+Our target for the first phase is Kubernetes v1.26, and from the end-users
+point of view, it is functionally equivalent to what is currently available in
+the container runtimes.
+
+#### Future phases
+
+It is important to think about the future steps to see the proposal in a
+broader context and understand what the full solution would look like.
+
+- Pod Spec: replace pod annotations with proper user interface in the PodSpec.
+- Resource discovery: a mechanism for the runtime to communicate available
+  QoS-class resource types and the available classes within each resource type
+  to kubelet
+- Resource status/capacity: hold the information of discovered resources (e.g.
+  in node status) for users to see what is available
+- Scheduler support: with resource discovery and status/capacity (above) in
+  place it is possible to handle QoS-class resource assignments in the
+  Kubernetes scheduler
+- Access control: mechanism to control which QoS-class resource types (and
+  classes within) are available for different users
+
+More details about the future steps are available in the [KEP][kep]
+(provisional).
+
+## Early access with container runtimes
+
+There is already some level of support for QoS-class resources in the container
+runtimes. Recent versions of CRI-O and containerd runtimes have support for
+configuring Intel RDT and blockio classes and controlling resource assignment
+via pod annotations. This allows early access and experimentation already
+before the Kubernetes support is there.
+
+Requirements:
+
+- Intel RDT
+  - CRI-O [v1.22][crio-release] (or later) or containerd
+    [v1.6][containerd-release] (or later)
+  - runc [v1.1][runc-release-11] (or later)
+  - hardware that supports that supports cache/memory bandwidth allocation
+    through the Linux resctrlfs interface
+- blockio
+  - CRI-O [v1.22][crio-release] (or later) or containerd mainline (no release
+    with blockio support is available, yet)
+  - runc [v1.0][runc-release-10] (or later)
+
+Pod annotations used for assigning the QoS-class resources:
+
+| Annotation  | Description
+| :---------- | :----------
+| `rdt.resources.beta.kubernetes.io/pod` | set a Pod-level default RDT class for all containers
+| `rdt.resources.beta.kubernetes.io/container.<container-name>` | set RDT class of one container
+| `blockio.resources.beta.kubernetes.io/pod` | set a Pod-level default blockio class for all containers
+| `blockio.resources.beta.kubernetes.io/container.<container-name>` | set blockio class of one container
```

Comment:
We can't blog about these until the KEP is accepted (and then, only after the annotations are [registered](https://kubernetes.io/docs/reference/labels-annotations-taints/))

I realize that Intel may have made some code that already uses these annotations (before anyone has registered them), but I'd prefer not to make it look like that use is official.

Reply:
These are annotations currently "understood" by the container runtimes. I pushed a change and tried to state it more clearly. I can remove the table if it's too much...

Reply:
> these are annotations currently "understood" by the container runtimes.

In an official blog, we absolutely should not mention these annotations. Either register them at the same time (and get that registration signed off), or don't use them.
Mostly that's a decision for SIG Architecture.

If some container runtimes pay attention to unregistered Kubernetes annotations - which would surprise me - then from a strict perspective those runtimes are welcome to. Let's not encourage other folks to take the same path though.

PS Bear in mind that having “alpha” or “beta” in annotation keys is deprecated / no longer done.

`rdt.resources.beta.kubernetes.io/pod` could instead be `rdt.something.intel.com/foo` and `blockio.resources.beta.kubernetes.io/pod` could be `linux-blockio.example/pod`. It's then much more clear that Kubernetes doesn't need to make any promise of special support.

Reply:
K, agree. I will rewrite these sections (whan/if the KEP gets accepted and this blog post can proceed)